tests/MongoDB.Driver.Tests/Specifications/unified-test-format/UnifiedValueMatcher.cs

```diff
@@ -0,0 +1,214 @@
+                                break;
+                            case "$$matchesHexBytes":
+                                // TODO: recheck
+                                expectedItem = specialOperator[0];
+                                break;
+                            default:
+                                throw new NotSupportedException($"Special operator not supported: '{specialOperator.GetElement(0).Name}'");
+                        }
+                    }
+
+                    actualDocument.Contains(expectedElement.Name).Should().BeTrue($"Actual document must contain key: {expectedElement.Name}");
+                    AssertValuesMatch(expectedItem, actualDocument[expectedElement.Name], false);
+                }
+
+                if (!isRoot)
+                {
+                    foreach (var name in actualDocument.Names)
+                    {
+                        expectedDocument.Contains(name).Should().BeTrue($"Actual document contains unexpected key: {name}");
+                    }
+                }
+            }
+            else if (expected.IsBsonArray)
+            {
+                actual.IsBsonArray.Should().BeTrue($"Actual value must be a document, but is '{actual.BsonType}'");
+                actual.AsBsonArray.Count.Should().Be(expected.AsBsonArray.Count, "Arrays must the be same size"); // TODO: Check if sizes are included in error message
+
+                var expectedArray = expected.AsBsonArray;
+                var actualArray = actual.AsBsonArray;
+
+                for (int i = 0; i < expectedArray.Count; i++)
+                {
+                    AssertValuesMatch(expectedArray[i], actualArray[i], false);
+                }
+            }
+            else if (expected.IsNumeric)
+            {
+                actual.IsNumeric.Should().BeTrue($"Actual value must be numeric, but is '{actual.BsonType}'");
+
+                expected.ToDouble().Should().Be(actual.ToDouble()); // TODO: Cast them to double? What about epsilon? Or maybe check each number type individually?
+            }
+            else
+            {
+                expected.BsonType.Should().Be(actual.BsonType);
+                expected.Should().Be(actual);
+            }
+        }
+
+        private void AssertExpectedType(BsonValue actual, BsonValue expectedTypes)
+        {
+            List<string> types;
+
+            if (expectedTypes.IsString)
+            {
+                types = new List<string> { expectedTypes.AsString };
+            }
+            else if (expectedTypes.IsBsonArray)
+            {
+                types = expectedTypes.AsBsonArray.Select(t => t.AsString).ToList();
+            }
+            else
+            {
+                throw new FormatException($"Unexpected $$type value BsonType: '{expectedTypes.BsonType}'");
+            }
+
+            types.Should().Contain(GetBsonTypeNameAsString(actual.BsonType));
```

Comment:
Actual and expected are reversed in this assertion. Should be:

    actualTypeName.Should().BeOneOf(expectedTypeNames);